### PR TITLE
Devirtualize PFBlockElement daughter classes

### PR DESCRIPTION
**Pull request by marksan87**
**PR description:**

This PR addresses the performance issues discussed in issue #33083. The `PFBlockElementTrack`, `PFBlockElementBrem`,  `PFBlockElementCluster`, `PFBlockElementSuperCluster`, and `PFBlockElementGsfTrack` daughter classes of `PFBlockElement` are devirtualized. Bitmasks are now used for `PFBlockElementGsfTrack::isSecondary` and in place of the logical operators used for `PFBlockElementTrack::isLinkedToDisplacedVertex`.

**PR validation:**

The performance was profiled with igprof using step 3 of matrix test 23434.21:

Base: https://msaunder.web.cern.ch/msaunder/cgi-bin/igprof-navigator/gitIssues/33083/step3_23434_21_base/1728
Final: https://msaunder.web.cern.ch/msaunder/cgi-bin/igprof-navigator/gitIssues/33083/step3_23434_21_final/3332

This is not a backport.

@hatakeyamak @bendavid @rappoccio @laurenhay

### CODE REVIEW

File: DataFormats/ParticleFlowReco/interface/PFBlockElementTrack.h [hunk 1]
```diff
@@ -50,12 +50,12 @@ namespace reco {
...
     /// check if the track is secondary
     bool isSecondary() const override {
-      return trackType(T_FROM_DISP) || trackType(T_FROM_GAMMACONV) || trackType(T_FROM_V0);
+      return trackType(T_FROM_DISP) | trackType(T_FROM_GAMMACONV) | trackType(T_FROM_V0);
     }
 
     bool isPrimary() const override { return trackType(T_TO_DISP); }
 
-    bool isLinkedToDisplacedVertex() const override { return isSecondary() || isPrimary(); }
+    bool isLinkedToDisplacedVertex() const override { return isSecondary() | isPrimary(); }
```
Comment (slava77) [inline DataFormats/ParticleFlowReco/interface/PFBlockElementTrack.h:58]: here and in `isPrimary` and `isSecondary` can we get bitmasks instead?

```suggestion
    static constexpr unsigned int kPrimaryMask = 1 << T_TO_DISP;
    static constexpr unsigned int kSecondaryMask = (1 << T_FROM_DISP) | (1 << T_FROM_GAMMACONV) | (1 << T_FROM_V0);
    static constexpr unsigned int kLinkedToDisplacedVertexMask = kPrimaryMask | kSecondaryMask;
    bool isSecondary() const override { return trackType_ & kSecondaryMask; } 
    bool isPrimary() const override { return trackType_ & kPrimaryMask; }
    bool isLinkedToDisplacedVertex() const override { return trackType_ & kLinkedToDisplacedVertexMask; }
```

it's worth testing the performance  though

Comment (zeratul87) [reply DataFormats/ParticleFlowReco/interface/PFBlockElementTrack.h]: Thanks for the suggestion! I have pushed these changes. The resulting performance improvement is non-negligible:

https://msaunder.web.cern.ch/msaunder/cgi-bin/igprof-navigator/gitIssues/33083/step3_23434_21_bitmasked/3400

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
Excellent. Would it make sense to have a look to all other daughters of "PFBlockElement"?

**Comment by slava77** [comment]
On 3/10/21 1:46 AM, Vincenzo Innocente wrote:
> Excellent. Would it make sense to have a look to all other daughters of
> "PFBlockElement"?
> 
> —

I agree, it makes sense to check there as well

**Comment by slava77** [comment]
> On 3/10/21 1:46 AM, Vincenzo Innocente wrote: Excellent. Would it make sense to have a look to all other daughters of "PFBlockElement"? —
> I agree, it makes sense to check there as well

OTOH, nothing else similarly relevant shows up in the profiler report https://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/releases/11_3_0_pre4/23434.21/step3/cpu/252

Perhaps a trivial finalization can still be done.

**Comment by zeratul87** [comment]
I have added devirtualization for the other PFBlockElement daughter classes here since it is a minor change.

**Comment by slava77** [comment]
@cmsbuild please test

@zeratul87 thank you for the quick update

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/33127	191b2d1
- other classes inheriting from `PFBlockElement` are now `final` 
- this is a technical update since the last reco signoff in https://github.com/cms-sw/cmssw/pull/33127#issuecomment-795694689 (no CPU performance changes are expected either)

**Comment by qliphy** [comment]
+1

**Pull request merged**